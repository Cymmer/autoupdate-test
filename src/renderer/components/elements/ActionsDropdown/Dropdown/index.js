import React from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';
import styles from './styles.module.scss';
import Card from '../../Card';

import actionTypes from '../constants/actionTypes';
import ActionLink from '../Action/Link';
import ActionButton from '../Action/Button';

const Dropdown = ({ isToggled, actions }) => (
  <Card
    className={cn({
      [styles.Dropdown]: !isToggled,
      [styles.Dropdown___toggled]: isToggled,
    })}
  >
    {actions?.map((action, index) =>
      action.type === actionTypes.LINK ? (
        <ActionLink
          key={`Action-${index}`}
          action={action}
          data-test="actionLink"
        />
      ) : (
        <ActionButton
          key={`Action-${index}`}
          action={action}
          data-test="actionButton"
        />
      )
    )}
  </Card>
);

Dropdown.propTypes = {
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      type: PropTypes.oneOf([actionTypes.BUTTON, actionTypes.LINK]),
      action: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
      icon: PropTypes.string,
      text: PropTypes.string,
      tooltip: PropTypes.string,
      disabled: PropTypes.bool,
    })
  ).isRequired,
  isToggled: PropTypes.bool.isRequired,
};

export default Dropdown;
