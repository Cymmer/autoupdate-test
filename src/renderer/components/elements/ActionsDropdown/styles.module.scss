.ActionsDropdown {
  position: relative;
  width: min-content;
}
