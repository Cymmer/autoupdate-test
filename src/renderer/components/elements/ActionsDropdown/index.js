import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from './styles.module.scss';

import IconButton from '../Button/IconButton';
import iconButtonTypes from '../Button/constants/iconButtonTypes';

import Dropdown from './Dropdown';
import actionTypes from './constants/actionTypes';

import { useOnClickOutside } from '../../../hooks';

const ActionsDropdown = ({ id, icon, iconButtonType, className, actions }) => {
  const ref = useRef();
  const [isDropdownToggled, toggleDropdown] = useState(false);
  useOnClickOutside(ref, () => toggleDropdown(false));

  if (actions.length === 0) {
    return null;
  }

  return (
    <div ref={ref} className={cn(className, styles.ActionsDropdown)}>
      <IconButton
        icon={icon}
        id={id}
        type={iconButtonType}
        onClick={() => toggleDropdown(!isDropdownToggled)}
      />
      <Dropdown actions={actions} isToggled={isDropdownToggled} />
    </div>
  );
};

ActionsDropdown.defaultProps = {
  className: null,
  icon: 'more_horiz',
  iconButtonType: iconButtonTypes.SOLID.LG,
  id: 'ellipseDropdown',
};

ActionsDropdown.propTypes = {
  // for dropdown container styling
  className: PropTypes.string,

  // decides which icon to use for dropdown toggle
  icon: PropTypes.string,

  // for mapping dropdown items
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      type: PropTypes.oneOf([actionTypes.BUTTON, actionTypes.LINK]),
      action: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
      icon: PropTypes.string,
      text: PropTypes.string,
      tooltip: PropTypes.string,
      disabled: PropTypes.bool,
    })
  ).isRequired,

  iconButtonType: PropTypes.oneOf([
    iconButtonTypes.SOLID.LG,
    iconButtonTypes.SOLID.MD,
    iconButtonTypes.SOLID.SM,
    iconButtonTypes.OUTLINE.LG,
    iconButtonTypes.OUTLINE.MD,
    iconButtonTypes.OUTLINE.SM,
  ]),

  id: PropTypes.string,
};

export default ActionsDropdown;
