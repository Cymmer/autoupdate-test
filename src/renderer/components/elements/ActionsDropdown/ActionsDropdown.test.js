import React from 'react';
import { shallow } from 'enzyme';
import ActionsDropdown from './index';
import actionTypes from './constants/actionTypes';

describe('RENDERING - ActionsDropdown', () => {
  it('should render with default props (empty actions)', () => {
    const component = shallow(<ActionsDropdown actions={[]} />);
    expect(component).toMatchSnapshot();
  });

  it('should render with actions', () => {
    const component = shallow(
      <ActionsDropdown
        actions={[
          {
            type: actionTypes.BUTTON,
            action: () => {},
            icon: 'test',
            text: 'Test',
            tooltip: 'Test Tooltip',
            disabled: false,
          },
          {
            type: actionTypes.BUTTON,
            action: () => {},
            icon: 'test',
            text: 'Test',
            tooltip: 'Test Tooltip',
            disabled: false,
          },
          {
            type: actionTypes.LINK,
            action: 'www.codechum.com',
            icon: 'test',
            text: 'Test',
            tooltip: 'Test Tooltip',
            disabled: false,
          },
          {
            type: actionTypes.LINK,
            action: 'www.codechum.com',
            icon: 'test',
            text: 'Test',
            tooltip: 'Test Tooltip',
            disabled: false,
          },
        ]}
        isToggled={false}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
