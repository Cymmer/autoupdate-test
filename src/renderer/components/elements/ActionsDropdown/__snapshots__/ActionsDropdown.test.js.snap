// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RENDERING - ActionsDropdown should render with actions 1`] = `
<div
  className="ActionsDropdown"
>
  <IconButton
    className={null}
    colorName="blue"
    disabled={false}
    icon="more_horiz"
    iconClassName={null}
    id="ellipseDropdown"
    isLocked={false}
    kind="button"
    onClick={[Function]}
    style={null}
    tabIndex={0}
    type="solid___lg"
  />
  <Dropdown
    actions={
      Array [
        Object {
          "action": [Function],
          "disabled": false,
          "icon": "test",
          "text": "Test",
          "tooltip": "Test Tooltip",
          "type": "button",
        },
        Object {
          "action": [Function],
          "disabled": false,
          "icon": "test",
          "text": "Test",
          "tooltip": "Test Tooltip",
          "type": "button",
        },
        Object {
          "action": "www.codechum.com",
          "disabled": false,
          "icon": "test",
          "text": "Test",
          "tooltip": "Test Tooltip",
          "type": "link",
        },
        Object {
          "action": "www.codechum.com",
          "disabled": false,
          "icon": "test",
          "text": "Test",
          "tooltip": "Test Tooltip",
          "type": "link",
        },
      ]
    }
    isToggled={false}
  />
</div>
`;

exports[`RENDERING - ActionsDropdown should render with default props (empty actions) 1`] = `""`;
